The first step is to read in the data using `read_csv`. We'll use an online file hosted by Github to provide the data, as local files can be tricky in Jupyter notebook.

In [ ]:
from urllib2 import urlopen
from pandas import read_csv
url = "https://raw.githubusercontent.com/amandamichelle27/IS362-Project1/master/flights.csv"
flights = read_csv(urlopen(url))
print flights

  Airline    Destination  On Time  Delayed
0  ALASKA    Los Angeles      497       62
1  ALASKA        Phoenix      221       12
2  ALASKA      San Diego      212       20
3  ALASKA  San Francisco      503      102
4  ALASKA        Seattle     1841      305
5  AMWEST    Los Angeles      694      117
6  AMWEST        Phoenix     4840      415
7  AMWEST      San Diego      383       65
8  AMWEST  San Francisco      320      129
9  AMWEST        Seattle      201       61

[10 rows x 4 columns]


As you can see above, I've formatted the data using one row cell in the original file, as this has the most flexibility. Specifically, each row provides the number of flights that are either on-time or delayed with a given destination for a given airline. For example, if I had instead used rows with enough columns for five different airports, I would have to add a whole new column if data was reported for another airport. Using this format will allow me to process additional data without changing my algorithm.

The first comparison I will perform is directly comparing the percentage of on-time flights versus total flights on a per-airport basis.

In [ ]:
df = flights  # Creating a copy for this section.
# Comput the percentage of online flights per destination and airline.
df["% On Time"] = 100 * df["On Time"] / (df["On Time"] + df["Delayed"])
# Drop the excess columns.
df = df.drop(["On Time", "Delayed"], 1)
# Create one table per airline and then join them by destination.
alaska = df[df["Airline"] == "ALASKA"].drop(["Airline"], 1)
amwest = df[df["Airline"] == "AMWEST"].drop(["Airline"], 1)
result = alaska.merge(amwest, on=["Destination"],
                      suffixes=(" ALASKA", " AMWEST"))
print result

     Destination  % On Time ALASKA  % On Time AMWEST
0    Los Angeles         88.908766         85.573366
1        Phoenix         94.849785         92.102759
2      San Diego         91.379310         85.491071
3  San Francisco         83.140496         71.269488
4        Seattle         85.787512         76.717557

[5 rows x 3 columns]


As you can see, Alaska has less delays proportionally for every airport. This proves that Alaska is the better airline based on the given data. Next, we'll examine the overall statistics per airline.

In [ ]:
df = flights  # Creating a copy for this section.
# Compute the total on-time and delayed flights per airline.
df = df.groupby(["Airline"])["On Time", "Delayed"].sum()
# Comput the percentage of on tine flights per destination and airline.
df["% On Time"] = 100 * df["On Time"] / (df["On Time"] + df["Delayed"])
# Drop the excess columns.
df = df.drop(["On Time", "Delayed"], 1)
print df

         % On Time
Airline           
ALASKA   86.728477
AMWEST   89.107266

[2 rows x 1 columns]


This result suggets that Amwest is the better airline, the opposite of our previous result. This is an example of Simpson's Paradox.<br><br>The best destination per airline is Phoenix for both of them. Amwest has significantly more flights traveling to Phoenix, so its overall result is nearly as good as just Phoenix (89.1% versus 92%). However, because Alaska rarely flies into Phoenix, its overall results is quite a bit worse than just Phoenix (86.7% versus 94.8%).<br><br>
If each airport had the same number of flights, then Alaska would have the better overall statistics. However, because the weights of each destination are different per destination, the overall results can be misleading. This goes to show that it's always good to perform an in-depth analysis before jumping to conclusions.